include/aws/common/string.h

```diff
@@ -0,0 +1,100 @@
+#ifndef AWS_COMMON_STRING_H
+#define AWS_COMMON_STRING_H
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/common/common.h>
+#include <aws/common/byte_buf.h>
+#include <string.h>
+
+/**
+ * Represents an immutable string holding either text or binary data. If the string is in constant
+ * memory or memory that should otherwise not be freed by this struct, set allocator to NULL
+ * and destroy function will be a no-op.
+ *
+ * This is for use cases where the entire struct and the data bytes themselves need to be
+ * held in dynamic memory, such as when held by a struct aws_hash_table. The data bytes
+ * themselves are always held in contiguous memory immediately after the end of the
+ * struct aws_string, and the memory for both the header and the data bytes
+ * is allocated together.
+ *
+ * Use the aws_string_bytes function to access the data bytes.
+ */
+struct aws_string {
+    struct aws_allocator * allocator;
+    size_t len;
+};
+
+static inline const uint8_t * aws_string_bytes(const struct aws_string * hdr) {
+    return (const uint8_t *)(hdr + 1);
+}
+
+static inline const struct aws_string * aws_string_from_literal_new(struct aws_allocator * allocator, const char * literal) {
+    size_t len = strlen(literal);
+    struct aws_string * hdr = aws_mem_acquire(allocator, sizeof(struct aws_string) + len);
+    if (!hdr) {aws_raise_error(AWS_ERROR_OOM); return NULL;}
+    hdr->allocator = allocator;
+    hdr->len = len;
+    memcpy((void *)aws_string_bytes(hdr), literal, len);
```

Comment:
This loses the null terminator. Should we preserve it (without reporting it as part of the length) - or, more in general, guarantee that `aws_string_bytes(hdr)[hdr->len] == 0`, as with C++ `std::string`? Doing so makes it much easier to interoperate with traditional C APIs.

Reply:
I feel strongly that we should not report the null terminator as part of the length. Not including it was consistent with our approach on byte buffers.

I don't have a huge objection to including it, but something to keep in mind is that the aws_string object is designed to hold either textual or binary data. In fact, the ability to use it to hold binary data is much more the justification for the need for the new data type in the context of a hash table. If we were only using null-terminated strings in the hash table, the pointers we pass in could be just pointers to the C-strings and we could always use strlen to determine their length. But this data type allows us to have keys and values within the hash table where the null byte is an ordinary character within the string. If we are treating the null byte as an ordinary character within the string, it feels a little unnatural to stick an extra null byte on the end. Furthermore, in that case getting a c-string back from the aws_string object would not give the expected result, because that c-string would end right before the first null byte.

Reply:
My argument any api we want to use that only takes a cstr argument without length is an API we shouldn’t be using in the first place.

Reply:
I don't think we should report it as part of the length. I'm simply considering that someone may reasonably want to take the value in an aws_string and pass it to, say, `printf` or `fopen` or something. Currently they'd need to malloc a new buffer, memcpy the data into it, set the nul-terminator, and then free the buffer. Or, at the cost of a single byte, we can make that problem go away.

Reply:
That's reasonable. I'll make the change to keep the extra null byte.